sentinel/src/schedule/outbound.js

```diff
@@ -0,0 +1,174 @@
+const objectPath = require('object-path'),
+      request = require('request-promise-native'),
+      vm = require('vm');
+
+const configService = require('../config'),
+      db = require('../db'),
+      logger = require('../lib/logger');
+
+const CONFIGURED_PUSHES = 'outbound';
+
+const queuedDocuments = () =>
+  db.medic.query('medic/outbound_queue', {include_docs: true})
+    .then(results => results.rows.map(r => r.doc));
+
+// Maps a source document to a destination format using the given push config
+const map = (doc, conf) => {
+  const srcParams = srcData => {
+    if (typeof srcData === 'string') {
+      return {
+        path: srcData,
+        required: true
+      };
+    } else {
+      return {
+        path: srcData.path,
+        expr: srcData.expr,
+        required: !srcData.optional
+      };
+    }
+  };
+
+  const toReturn = {};
+
+  Object.keys(conf.mapping).forEach(dest => {
+    const {
+      path, expr, required
+    } = srcParams(conf.mapping[dest]);
+
+    let srcValue;
+
+    if (expr) {
+      try {
+        srcValue = vm.runInNewContext(expr, {doc: doc});
+      } catch (err) {
+        throw Error(`Mapping error for ${conf.name}:${dest}: JS error on source document: ${doc._id}: ${err}`);
+      }
+    } else {
+      srcValue = objectPath.get(doc, path);
+    }
+
+    if (required && !srcValue) {
+      throw Error(`Mapping error for ${conf.name}:${dest}: no source present`);
+    }
+
+    if (srcValue) {
+      objectPath.set(toReturn, dest, srcValue);
+    }
+  });
+
+  return toReturn;
+};
+
+// Attempts to send a given payload using a given push config
+const send = (payload, conf) => {
+  const sendOptions = {
+    method: 'POST',
+    url: conf.destination.base_url + conf.destination.path,
+    body: payload,
+    json: true
+  };
+
+  const auth = () => {
+    const authConf = conf.destination.auth;
+
+    if (!authConf) {
+      return Promise.resolve();
+    }
+
+    if (authConf.type === 'Basic') {
+      sendOptions.auth = {
+        username: authConf.username,
+        password: authConf.password,
+        sendImmediately: true
+      };
+      return Promise.resolve();
+    }
+
+    if (authConf.type === 'Muso') {
+      const authOptions = {
+        method: 'POST',
+        form: {
+          login: authConf.username,
+          password: authConf.password
+        },
+        // TODO: I think Request adds this for us?
+        // headers: {
+        //   'Content-Type': 'multipart/form-data'
+        // },
+        url: conf.destination.base_url + authConf.path
+      };
+
+      return request(authOptions)
```

Comment:
Have you considered using the `async` pattern when working with promises in new code in api and sentinel?

Reply:
I **completely forgot** we can use that now!

We _can_ use that now right?

If so I'll rewrite, it's a good self-contained situation to play with it.

Reply:
I forgot to write an update here: we don't support async / await without also changing our eslint config. Which is no problem, but I have too much on my plate to justify rewriting existing working code just to change patterns, so I'll make that change next time and write async from scratch